### ISSUE TITLE
Compatible to Redmine 4.x

### ISSUE THREAD

**Issue by oxyii** [jbox-web/redmine_git_hosting#725]
Now we planed to upgrade from Redmine 3.4.6 (stable) to the newer version 4.0. Actually we use the plugin version 1.2.3. Is this compatible to Redmine 4.x?

**Comment by andrenasturas** [comment]
No it isn't, unfortunately.
I tried an upgrade to Redmine 4.0.0, with the ideal requierements Ruby 2.5 and Rails 5.2. I had to resolve some dependencies conflicts in the plugin's Gemfile : 
```
-	gem 'gitlab-grack', git: 'https://github.com/jbox-web/grack.git', require: 'grack', branch: 'fix_rails4'
-	gem 'redcarpet': '~> 3.3.2'
+	gem 'redcarpet': '~> 3.4'
+	gem 'coderay'
```
But then the database migration still fails.
```
rake aborted!
NoMethodError: undefined method `alias_method_chain' for RepositoriesController:Class
Did you mean?  alias_method
/home/redmine/redmine-4.0.0/plugins/redmine_git_hosting/lib/redmine_git_hosting/patches/repositories_controller_patch.rb:13:in `block in included'
...
```
The plugin uses a method called `alias_method_chain`, which I found [here](https://github.com/mikitex70/redmine_drawio/issues/44) to be deprecated with Rails 5. Since Redmine 4 uses Rails 5.2, we have here a problem. As I tried some no-brain brute-force like patching, I found other methods unsupported by Rails 5, like `before_filter` and `skip_before_filter`.

I know almost nothing about Ruby, so that's here my investigation ends. I hope that an upgrade will be released soon, I really like that plugin and I use it a lot :)

**Comment by jcspader** [comment]
+1

**Comment by Alwaysin** [comment]
+1 waiting for this to migrate!

**Comment by jserve** [comment]
I am just curios, is someone working for being compatible with Redmine 4?

**Comment by PowerKiKi** [comment]
Some attempts were made in #727, but unfortunately it was not successful.

@n-rodriguez, could you please comment on the situation ? it is not clear if you are still actively maintaining this project or not ? would you be able to accept a PR for Redmine 4 support ?

**Comment by oxyii** [comment]
The plugin partially lost its meaning after github was making private repo free... Additionally, looks like the project is dead... :disappointed: I migrated to gitlab first (very comfortable with the docker by the way) but now I’ve got enough github functionality. Good luck to all!

**Comment by mrPsycho** [comment]
that sad.

this plugin was great...

**Comment by oxyii** [comment]
@mrPsycho 

> this plugin was great...

I certainly agree. I put it on all dev-servers about 5 years and actively used. I even wrote a bash-script for integration with ispManager. It was a miserable kind of CD, when this term was not even used. VERY sorry! I'm out (

**Comment by n-rodriguez** [comment]
> looks like the project is dead... disappointed

not totally

**Comment by PowerKiKi** [comment]
Redmine has a lot of features that just does not exist in GitHub (mostly for project managing oriented stuff). And this plugins is great to fill the gap of code hosting where Redmine does not shine as much as it could. I am convinced that Redmine and this plugin still have very good reasons to exist.

@n-rodriguez could you please elaborate on your situation, so the community can organize itself accordingly ? do you plan to work on Redmine 4 support yourself ? or would you prefer to merge a PR submitted by the community ? or should we just fork this repository ?

**Comment by Llewellynvdm** [comment]
This is urgent, and we would be willing to give donations....

**Comment by shugg** [comment]
Any update on this topic?

**Comment by pva** [comment]
Looks like these guys started to work on this: https://github.com/AlphaNodes/redmine_git_hosting See, there is branch 4.0-stable.

**Comment by n-rodriguez** [comment]
> Looks like these guys started to work on this: https://github.com/AlphaNodes/redmine_git_hosting See, there is branch 4.0-stable.

I'm looking for maintainers. Mail me if you're interested. It would be great to not fork again.

**Comment by PowerKiKi** [comment]
@alexandermeindl, since you've been working on #727 and are actively working on https://github.com/AlphaNodes/redmine_git_hosting would you be interested in joining this project as a maintainer as suggested by @n-rodriguez ? or would you rather continue on your own fork ?

**Comment by alexandermeindl** [comment]
=== Github Anwort git hosting

https://github.com/jbox-web/redmine_git_hosting/issues/725

Hi @PowerKiKi,
the problem is, that my PR did not work with [autoloading](https://guides.rubyonrails.org/autoloading_and_reloading_constants.html) in production. Now I am working in my branch on a working solution for Redmine 4. I used the devel branch as starting point of my work and I've to figure out, what I have to fix to get it stable. There is lots of work to do.
I'll make a PR, if it is working. So in my ideal world, my fork as a new project is not required. But I changed/change a lot and maybe my solutions are not compatible with the idea of @n-rodriguez.
My main goal now: to get a running solution for Redmine 4

@n-rodriguez did an absolute great job and I can image, how many hours and passion he spent on this project. 1000 thanks for that, man!
Git hosting is a huge topic and in my opinion, we can only get it for Redmine as an current day solution together. This means, Redmine community and their involved companies. That's not the job of a One-Man-Show. If you compare it to Gitlab, they have an active development with hundreds of commits almost daily - with venture capital.

**Comment by n-rodriguez** [comment]
@alexandermeindl thanks a lot for your hard work :+1: . I've added you as a collaborator on the projet. You should have write/merge permissions. Feel free to submit your PR when it's ready so we can discuss about it but you'll be able to merge it.

> But I changed/change a lot and maybe my solutions are not compatible with the idea of @n-rodriguez. My main goal now: to get a running solution for Redmine 4

So we're in tune.

**Comment by PowerKiKi** [comment]
@n-rodriguez since #727 got merged I couldn't see much activity here. Could you please share your point of view on what is still needed to release a Redmine 4 compatible version ? and who might be working on what ? or whether help is needed for some things ?

**Comment by Alwaysin** [comment]
@PowerKiKi FYI, @alexandermeindl 's fork situated here: https://github.com/AlphaNodes/redmine_git_hosting is highly functional, we use it in production with latest Redmine, 4.0.4.

**Comment by n-rodriguez** [comment]
@PowerKiKi FYI @alexandermeindl code has been merged in this repo.

**Comment by PowerKiKi** [comment]
@n-rodriguez I saw that it was merged, but it was never released and this issue is still open. So that seems to indicate that there are still things that you plan to do. But you didn't answer my questions as to what is left to do and who might be working on it ?

@Alwaysin thank you for reporting your experience. I'll give it a try in the coming days

**Comment by Alwaysin** [comment]
@alexandermeindl has protected branches functionality been removed on purpose?
ALso is it aybe possible to enable issues module of github on your fork?

**Comment by alexandermeindl** [comment]
Hi @Alwaysin,

protected branches are not removed by me. It it doesn't work, it is a bug. Please report a bug as new issue in this repo, because my fork is not longer maintained and it'll be removed soon. Development/bug fixing should be here in this repo.

**Comment by edmcdonagh** [comment]
@alexandermeindl, @n-rodriguez - do you know when a release might be made?

**Comment by Alwaysin** [comment]
@edmcdonagh FYI you can already use the code without a "release", which is not really needed.

> git clone https://github.com/jbox-web/redmine_git_hosting.git

is enough

**Comment by PowerKiKi** [comment]
We've been using devel branch in production for a few days with success. But a release would be nice, if only to formally announce Redmine 4 compatibility. Specifically master branch is now severely outdated.

**Comment by n-rodriguez** [comment]
> We've been using devel branch in production for a few days with success. But a release would be nice, if only to formally announce Redmine 4 compatibility. Specifically master branch is now severely outdated.

I do agree. I need to find the time to do it. I'm really sorry for the slow development in the last 2 or 3 years.

**Comment by PowerKiKi** [comment]
redmine_git_hosting 4.0.0 has just been released and we now support (only) Redmine 4.0.3+.
